brave/brave-core
# Migrate crypto widgets visibility in NTP

fix https://github.com/brave/brave-browser/issues/19708

Migration is done in the first NTP loading only once.

For new users, all crypto widgets are hidden by default.
For existing users, if current foremost widget is crypto widget,
only it is visible and others are hidden.
After migration, user can control crypto widget's visibility from gallery.
    
To make all crypto widgets hidden by default, each pref's default
value is not changed. Their default prefs is still "true".
As their visibility are managed by prefs and widget order is managed in local
storage, it's difficult to start migration from c++.

If user has already authed crypto widgets, they will be also visible regardless of position in the widget stack.


Resolves 

## Submitter Checklist:


## Reviewer Checklist:


## After-merge Checklist:

  changes has landed on

## Test Plan:

### Test 1 (New user)
1. Launch with clean profile
2. Check all crypto widgets are hidden in NTP widget stack
3. Control their visibility from gallery and check visibility are retained with new NTP or after restarting

### Test 2-1 (Existing user)
1. Launch old version brave with existing profile
2. Show some(or all) crypto widgets in stack but set Rewards as a foremost
3. Close old version brave and Launch new Brave
4. Check all crypto widgets are hidden in widget stack

### Test 2-2 (Existing user)
1. Launch old version brave with existing profile
2. Show some(or all) crypto widgets in stack and set one of crypto widget as a foremost one
3. Close old version brave and Launch new Brave
4. Check only foremost crypto widget is visible in the stack

### Test 2-3 (Existing user that has authed widget)
1. Launch old version brave with existing profile
2. Connect ftx widget
3. Show some(or all) crypto widgets in stack and set one of crypto widget
4. Close old verion brave and Launch new Brave
5. Check ftx widget is not hidden regardless of its position
